### PR TITLE
Add generator and worker seed

### LINKED ISSUE #8601: Reproducibility in multi-process data loading
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

[In the official documentation](https://pytorch.org/docs/stable/notes/randomness.html#dataloader) it is recommended to set the rseed for the worker to ensure reproducibility even for multi-process dataloaders.

### Use case

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by UnglvKitDe**
Worker seed and generator inserted into the dataloader as described in #8601

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhance randomness control in YOLOv5 dataloaders for more consistent training results. 🔄

### 📊 Key Changes
- Added `seed_worker` function to initialize the random seeds for dataloader workers.
- Incorporated setting of manual seed and generator for PyTorch DataLoader.

### 🎯 Purpose & Impact
- 🎲 **Improve Consistency**: Ensure that each worker in a dataloader initializes with a specific seed for better reproducibility.
- ✨ **Enhanced Randomness**: The changes allow for more controlled randomness during data loading, which can contribute to more stable training performance across different runs.
- 👩‍🔬 **Research Friendly**: Facilitate experimental consistency for researchers and developers, enabling fair comparison of models and training regimes.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@UnglvKitDe I'm seeing identical reproducible results with master and torch>=1.12.0. It doesn't seem we need any more modifications. i.e. in Colab with `python train.py --epochs 3`:
Screen Shot 2022-07-19 at 6 21 53 PM

**Comment by glenn-jocher** [comment]
@UnglvKitDe are these DDP-specific requirements for reproducibility?

**Comment by UnglvKitDe** [comment]
@glenn-jocher Mh, so I implemented it in my version because it was recommended and I didn't want to take any risks with edge cases. On the other hand, because I once read about a bug like [this](https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects/). But it seems that it has been fixed in the meantime. I could not recreate the case, but it is also still described in tutorials like [here ](https://towardsdatascience.com/random-seeds-and-reproducibility-933da79446e3)that it can come to this problem especially with many workers.

**Comment by glenn-jocher** [comment]
@UnglvKitDe I can verify that I am not seeing reproducible results with DDP trainings, we probably do need this PR merged then.

I'm worried about the same seed on all workers though, I think this may impact augmentation by repeating the same augmentations. Perhaps we should set worker seeds equal to their RANK?

**Comment by glenn-jocher** [comment]
@UnglvKitDe PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Comment by glenn-jocher** [comment]
@UnglvKitDe tested DDP training but I do not see reproducible results after this PR unfortunately. It seems something else is missing.

**Comment by UnglvKitDe** [comment]
> @UnglvKitDe I can verify that I am not seeing reproducible results with DDP trainings, we probably do need this PR merged then.
> 
> I'm worried about the same seed on all workers though, I think this may impact augmentation by repeating the same augmentations. Perhaps we should set worker seeds equal to their RANK?

@glenn-jocher Each worker has a different worker_seed. Here is a small example:
```
class RandomDataset(Dataset):
    def __getitem__(self, index):
        return np.random.randint(0, 1, 1)

    def __len__(self):
        return 16

def seed_worker(worker_id):
    # https://pytorch.org/docs/stable/notes/randomness.html#avoiding-nondeterministic-algorithms
    init_seed = torch.initial_seed()
    worker_seed = init_seed % 2**32
    print(os.getpid(), worker_seed, init_seed)
    np.random.seed(worker_seed)
    random.seed(worker_seed)
dataset = RandomDataset()
dataloader = DataLoader(dataset, batch_size=2, num_workers=4, 
                        worker_init_fn=seed_worker)
for epoch in range(1):
    print(f"epoch: {epoch}")
    for batch in dataloader:
        print(batch)
    print("-"*25)
```
image

**Comment by glenn-jocher** [comment]
@UnglvKitDe oh perfect, got it! I tried to implement an additional PR https://github.com/ultralytics/yolov5/pull/8688 but still don't get reproducible DDP results even afterwards. Not sure unfortunately.
